fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java

```diff
@@ -270,38 +270,39 @@ private LogicalPlan processSetOperation(SetOperationRelation setOperationRelatio
         Scope outputScope = setOperationRelation.getRelations().get(0).getScope();
         ExpressionMapping expressionMapping = new ExpressionMapping(outputScope, outputColumns);
 
-        LogicalOperator setOperator;
-        OptExprBuilder subOpt;
+        OptExprBuilder root;
         if (setOperationRelation instanceof UnionRelation) {
-            setOperator = new LogicalUnionOperator.Builder()
+            root = new OptExprBuilder(new LogicalUnionOperator.Builder()
                     .setOutputColumnRefOp(outputColumns)
                     .setChildOutputColumns(childOutputColumnList)
                     .isUnionAll(!SetQualifier.DISTINCT.equals(setOperationRelation.getQualifier()))
-                    .build();
+                    .build(),
+                    childPlan, expressionMapping);
 
             if (setOperationRelation.getQualifier().equals(SetQualifier.DISTINCT)) {
-                OptExprBuilder unionOpt = new OptExprBuilder(setOperator, childPlan, expressionMapping);
-                subOpt = new OptExprBuilder(
-                        new LogicalAggregationOperator(AggType.GLOBAL, outputColumns, Maps.newHashMap()),
-                        Lists.newArrayList(unionOpt), expressionMapping);
-            } else {
-                subOpt = new OptExprBuilder(setOperator, childPlan, expressionMapping);
+                root = root.withNewRoot(new LogicalAggregationOperator(AggType.GLOBAL, outputColumns, Maps.newHashMap()));
             }
         } else if (setOperationRelation instanceof ExceptRelation) {
-            setOperator = new LogicalExceptOperator.Builder()
+            root = new OptExprBuilder(new LogicalExceptOperator.Builder()
                     .setOutputColumnRefOp(outputColumns)
-                    .setChildOutputColumns(childOutputColumnList).build();
-            subOpt = new OptExprBuilder(setOperator, childPlan, expressionMapping);
+                    .setChildOutputColumns(childOutputColumnList).build(),
+                    childPlan, expressionMapping);
+
         } else if (setOperationRelation instanceof IntersectRelation) {
-            setOperator = new LogicalIntersectOperator.Builder()
+            root = new OptExprBuilder(new LogicalIntersectOperator.Builder()
                     .setOutputColumnRefOp(outputColumns)
-                    .setChildOutputColumns(childOutputColumnList).build();
-            subOpt = new OptExprBuilder(setOperator, childPlan, expressionMapping);
+                    .setChildOutputColumns(childOutputColumnList).build(),
+                    childPlan, expressionMapping);
         } else {
             throw unsupportedException("New Planner only support Query Statement");
         }
 
-        return new LogicalPlan(subOpt, outputColumns, null);
+        LimitElement limit = setOperationRelation.getLimit();
```

Comment:
we do not deal with union + limit before?

Reply:
The previous limit was assigned to select, which was a wrong approach. Should be assigned to SetOp